cpp/doxygen/developer_guide/DEVELOPER_GUIDE.md

```diff
@@ -901,6 +901,42 @@ void trivial_types_only(T t){
 }
 ```
 
+# Logging
+
+libcudf includes logging utilities (built on top of [spdlog](https://github.com/gabime/spdlog)
+library), which should be used to log important events (e.g. user warnings). This utility can also
+be used to log debug information, as long as the correct logging level is used. There are six macros
+that should be used for logging at different levels:
+
+* `CUDF_LOG_TRACE` - verbose debug messages (targeted at developers)
+* `CUDF_LOG_DEBUG` - debug messages (targeted at developers)
+* `CUDF_LOG_INFO` - information about rare events (e.g. once per run) that occur during normal
+execution
+* `CUDF_LOG_WARN` - user warnings about potentially unexpected behavior or deprecations
+* `CUDF_LOG_ERROR` - recoverable errors
+* `CUDF_LOG_CRITICAL` - unrecoverable errors (e.g. memory corruption)
+
+By default, messages below the `INFO` level are excluded from the log. In public builds, the code
+that logs at these levels is compiled out. This prevents logging of potentially sensitive data that
+might be done for debug purposes. Also, this allows developers to include expensive computation in
+the trace/debug logs, as the overhead will not be present in the public builds. The minimum enabled
+logging level is `INFO`, and it can be modified in multiple ways:
+
+* CMake configuration variable `LIBCUDF_LOGGING_LEVEL` - controls the minimum level of logging that
+will be compiled in the build.
+Available levels are `TRACE`, `DEBUG`, `INFO`, `WARN`, `ERROR`, `CRITICAL`, and `OFF`.
+* Environment variable `LIBCUDF_LOGGING_LEVEL`, with the same values as the CMake variable.
+The difference is that disabled levels are present in the build, but the messages are excluded from
+the log.
+* Global logger object exposed via `cudf::logger()`. This API should not be used within libcudf to
+manipulate logging,
+its purpose is to allow upstream users to configure libcudf logging to fit their application.
+
+The default log file is `cudf_log.txt` in the current working directory.
+The environment variable `LIBCUDF_DEBUG_LOG_FILE` can be set to specify the path and file name.
+Upstream users can also manipulate `cudf::logger().sinks()` to add sinks or divert the log to standard
+output or even a custom spdlog sink.
```

Comment:
I would recommend adding support to have different sinks for different levels. By doing so, we can leverage multiple output log streams for debugging/application monitoring etc.

Reply:
I would argue that's not necessary.  If someone wants to do that, they can add a sink that multiplexes out to the separate streams based not just on log level but any other custom logic they might want to use to determine which stream to use.

Reply:
Example of multi-sinks for multi-levels from spdlog: https://github.com/gabime/spdlog/blob/3cab260814613b0a7e6889d2a030b2992c916d1a/example/example.cpp#L253

Reply:
Oh wait. We may support that implicitly via spdlog! For example, we support doing so (in the example below):
```
cudf::logger().sinks().push_back(std::make_shared<spdlog::sinks::stdout_sink_mt>());
```

And the example from spdlog has a snippet to create a sink like this:
```
auto file_sink = std::make_shared<spdlog::sinks::basic_file_sink_mt>("logs/multisink.txt", true);
```

So we can achieve multi-sinks for multi-levels by just setting the corresponding level for each sink. Please help verify this and update this dev guide too.